src_c/surface.c

```diff
@@ -2105,6 +2116,313 @@ surf_blits(pgSurfaceObject *self, PyObject *args, PyObject *keywds)
     return RAISE(PyExc_TypeError, "Unknown error");
 }
 
+#if PY_VERSION_HEX >= 0x03070000
+#define UBLITS_ERR_TUPLE_REQUIRED 11
+#define UBLITS_ERR_INSUFFICIENT_ARGS 12
+#define UBLITS_ERR_FLAG_NOT_NUMERIC 13
+#define UBLITS_ERR_DORETURN_NOT_NUMERIC 14
+#define UBLITS_SERR_LISTORTUPLE_REQUIRED \
+    "blit_sequence can only be a list or a tuple"
+static PyObject *
+surf_ublits(pgSurfaceObject *self, PyObject *const *args, Py_ssize_t nargs)
+{
+    SDL_Surface *src, *dest = pgSurface_AsSurface(self);
+    SDL_Rect *src_rect, temp, dest_rect;
+
+    PyObject *srcobject = NULL, *argpos = NULL;
+    PyObject *blitsequence, *tmpblitseq;
+    PyObject *iterator = NULL;
+    PyObject *item = NULL;
+    PyObject *ret = NULL;
+    PyObject *retrect = NULL;
+    PyObject **f_blitsequence;
+    Py_ssize_t itemlength, sequencelength, seq_counter;
+    int doreturn = 1;
+    int errornum = 0;
+    int result;
+    int flags_numeric;
+
+    temp.x = 0;
+    temp.y = 0;
+
+    if (nargs < 3) {
+        errornum = UBLITS_ERR_INSUFFICIENT_ARGS;
+        goto on_error;
+    }
+
+    blitsequence = args[0];
+    if (!PyLong_Check(args[1]) || !pg_IntFromObj(args[1], &flags_numeric)) {
+        errornum = UBLITS_ERR_FLAG_NOT_NUMERIC;
+        goto on_error;
+    }
+
+    if (!pg_IntFromObj(args[2], &doreturn)) {
+        errornum = UBLITS_ERR_DORETURN_NOT_NUMERIC;
+        goto on_error;
+    }
+
+    if (!dest) {
+        errornum = BLITS_ERR_DISPLAY_SURF_QUIT;
+        goto on_error;
+    }
+
+    /* Generator */
+    if (PyGen_Check(blitsequence)) {
+        if (doreturn) {
+            ret = PyList_New(0);
+            if (!ret)
+                return NULL;
+        }
+        if (!(PyIter_Check(blitsequence))) {
+            errornum = BLITS_ERR_SEQUENCE_REQUIRED;
+            goto on_error;
+        }
+        iterator = PyObject_GetIter(blitsequence);
+        if (!iterator) {
+            Py_XDECREF(ret);
+            return NULL;
+        }
```

Comment:
The type checking -> execution paths in this PR make me nervous.

This pseudocode is how it works now
```
if generator:
    if iterator:  # <- what generator wouldn't be an iterator?
       PyIter strategy
    else:
        fail

else if sequence:
    if sequence_fast_support: # <- this is not the same as if sequence, there are more sequences than lists or tuples
        sequence_fast strategy
    else:
        fail

else:
    fail
```

I think it should be something like
```
if sequence_fast_support:
    sequence_fast strategy
else if itererator:
    iter strategy
else
    fail

Reply:
And by if sequence_fast_support, I just mean checking whether `PySequence_Fast` returned NULL. (Also you can use PyErr_Clear to clear the resulting set error if it is NULL).

Reply:
Yeah i see the problem... but PySequence_Fast somehow corrupts the generator, invalidating the iteration

Reply:
if you put it before the generator

Reply:
i'll tackle this tomorrow

Reply:
if you do this 
```
if PySequence_Fast(blitsequence, errmsg): <- is always true with generators/lists/tuples
    sequence_fast strategy
else if PyIter_Check(blitsequence):
    iter strategy
else
    fail
```

best i can do is 
```
if PySequence_Fast(blitsequence, errmsg) and !PyGen_Check(blitsequence):
    sequence_fast strategy
else if PyIter_Check(blitsequence):
    iter strategy
else
    fail
```

Reply:
mosty importantly i noticed this function segfaults if strings are passed as blit_sequence

Reply:
> Would lead to too much nesting because at that point you have to check for each type inside the first if clause and implement each strategy all inside that block. Not to say you'd have to DECREF nmultiple times if the type doesn't exactly match a list or tuple as calling PySequence_Fast includes assigning its return to a variable.

- For strings, check that before and exit on wrong type.
- "implement each strategy in each block" -> no, if it enters the first block, use sequence_fast.
- "you'd have to DECREF nmultiple times" -> no, if it gives you a valid reference, use it, otherwise it will give you NULL.